backtesting/backtesting.py

```diff
@@ -1221,101 +1236,172 @@ def maximize(stats: pd.Series, _key=maximize):
-                for batch_index in _tqdm(range(len(param_batches))):
-                    _, values = Backtest._mp_task(backtest_uuid, batch_index)
-                    for value, params in zip(values, param_batches[batch_index]):
-                        heatmap[tuple(params.values())] = value
-        finally:
-            del Backtest._mp_backtests[backtest_uuid]
-
-        best_params = heatmap.idxmax()
-
-        if pd.isnull(best_params):
-            # No trade was made in any of the runs. Just make a random
-            # run so we get some, if empty, results
-            self.run(**param_combos[0])  # type: ignore
-        else:
-            # Re-run best strategy so that the next .plot() call will render it
-            self.run(**dict(zip(heatmap.index.names, best_params)))
+            finally:
+                del Backtest._mp_backtests[backtest_uuid]
 
-        if return_heatmap:
-            return self._results, heatmap
-        return self._results
+            best_params = heatmap.idxmax()
 
-    @staticmethod
+            if pd.isnull(best_params):
+                # No trade was made in any of the runs. Just make a random
+                # run so we get some, if empty, results
+                self.run(**param_combos[0])  # type: ignore
+            else:
+                # Re-run best strategy so that the next .plot() call will render it
+                self.run(**dict(zip(heatmap.index.names, best_params)))
+
+            if return_heatmap:
+                return self._results, heatmap
+            return self._results
+
+        def _optimize_skopt() -> Union[pd.Series, Tuple[pd.Series, pd.Series]]:
+            """
+            Optimize strategy parameters to an optimal combination using
+            scikit-optimize. Returns result `pd.Series` of
+            the best run.
+
+            `maximize` is a string key from the
+            `backtesting.backtesting.Backtest.run`-returned results series,
+            or a function that accepts this series object and returns a number;
+            the higher the better. By default, the method maximizes
+            Van Tharp's [System Quality Number](https://google.com/search?q=System+Quality+Number).
+
+            Additional keyword arguments represent strategy arguments with
+            list-like collections of possible values. For example, the following
+            code finds and returns the "best" of the 7 admissible (of the
+            9 possible) parameter combinations:
+
+                backtest.optimize(sma1=[5, 10, 15], sma2=[10, 20, 40],
+                                constraint=lambda p: p.sma1 < p.sma2)
+
+            .. TODO::
+                Improve multiprocessing with joblib.parallel .
+            """
+            dimensions = []
+            for key, value in kwargs.items():
+                args_array = np.array(value)
+
+                if args_array.dtype.kind == 'i':
```

Comment:
[`kind in 'ium'`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html)